File: typescript/runtime/src/SerializationFormat.ts
import {DeserializationStrategy, SerializationStrategy} from "./serialization/TsSerializer";

export interface SerializationFormat {
    encode<T>(serializer: SerializationStrategy<T>, value: T): Uint8Array
    decode<T>(deserializer: DeserializationStrategy<T>, raw: Uint8Array): T
}

File: kotlin/lib/src/commonMain/kotlin/com/caesarealabs/rpc4k/runtime/api/SerializationFormat.kt
package com.caesarealabs.rpc4k.runtime.api
import com.caesarealabs.rpc4k.runtime.implementation.serializers.*
import io.ktor.http.*
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.SerializationStrategy
import kotlinx.serialization.json.Json


/**
 * For example [Json] or Protobuf
 * Must use the [Rpc4kSerializersModule]
 */
public interface SerializationFormat {
    public fun <T> encode(serializer: SerializationStrategy<T>, value: T): ByteArray
    public fun <T> decode(serializer: DeserializationStrategy<T>, raw: ByteArray): T
    public val contentType: ContentType
}

